File: data_structures/cpp/stack_test.cpp
#define CATCH_CONFIG_MAIN
#include "catch.hpp"
#include "stack.hpp"
#include <iostream>

using namespace std;


TEST_CASE("Stack Works Completely", "[Stack]") {
    cout << "Hello World" << endl;

    Stack<int> stack;

    REQUIRE(stack.IsEmpty() == true);

    stack.Push(2);
    stack.Push(5);
    stack.Push(12);
    stack.Push(40);

    REQUIRE(stack.IsEmpty() == false);
    REQUIRE(stack.Size() == 4);

    REQUIRE(stack.Peek() == 40);
    REQUIRE(stack.Pop() == 40);
    REQUIRE(stack.Pop() == 12);

    REQUIRE(stack.Size() == 2);
    REQUIRE(stack.Peek() == 5);
    REQUIRE(stack.Pop() == 5);
    REQUIRE(stack.Pop() == 2);

    REQUIRE(stack.IsEmpty() == true);

    stack.Push(12);
    stack.Push(40);
}

File: data_structures/go/stack_test.go
package ds

import (
	"testing"

	"gotest.tools/assert"
)

func TestStack(t *testing.T) {
	stack := Stack{}
	assert.Assert(t, stack.IsEmpty())

	stack.Push(5)
	stack.Push(2)
	stack.Push(19)
	stack.Push(48)

	assert.Assert(t, !stack.IsEmpty())

	assert.Equal(t, 4, stack.Size())
	assert.Equal(t, 48, stack.Peek())
	assert.Equal(t, 48, stack.Pop())
	assert.Equal(t, 19, stack.Pop())
	assert.Equal(t, 2, stack.Size())

	assert.Equal(t, 2, stack.Peek())
	assert.Equal(t, 2, stack.Pop())
	assert.Equal(t, 5, stack.Pop())

	assert.Assert(t, stack.IsEmpty())

}
